### PR TITLE
New resource r/aws_wafv2_rule_group

### LINKED ISSUE #11175: Feature Request: WAFv2 Rule Group Resource
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

In November 2019, AWS released a new version of the WAF API, [WAFv2](https://docs.aws.amazon.com/waf/latest/APIReference/Welcome.html), which offers improved functionality over the previous WAF API ("WAF Classic") such as [Managed Rules](https://github.com/terraform-providers/terraform-provider-aws/issues/11046) and WAF Capacity Units. This new API requires separate Terraform resource implementations from the previous resource implementations.

### New or Affected Resource(s)

* aws_wafv2_rule_group

### Potential Terraform Configuration

```hcl
resource "aws_wafv2_rule_group" "example" {
  capacity          = 1
  description        = ""
  name               = ""
  scope              = ""
  tags               = {}

  rule {
    action {
      # The below shows the current API structure, which means there may be intention to include
      # additional settings with each action type in the future
      # Purposefully empty configuration blocks are awkward in Terraform configurations and the Plugin SDK though, so we may want to shy away from the pure API schema mapping here if they cause any trouble
      allow {}
      block {}
      count {}
    }

    name = ""
    
    override_action {
      count {}
      none {}
    }

    priority = 1

    # these are not fully shown and appear to be recursive in some cases    
    statement {
      and_statement {
        statement {}
        statement {}
      }

      byte_match_statement {
        field_to_match {
          all_query_arguments {}
          body {}
          method {}
          query_string {}
          single_header {
            name = ""
          }
          single_query_argument {
            name = ""
          }
          uri_path {}
        }

        positional_constraint = ""
        search_string = ""

        text_transformation {
          priority = 1
          type = ""
        }
      }

      geo_match_statement {
        country_codes = []
      }

      ip_set_reference_statement {
        arn = ""
      }

      managed_rule_group_statement {
        excluded_rule {
          name = ""
        }

        name = ""
        vendor_name = ""
      }

      not_statement {
        statement {}
      }

      or_statement {
        statement {}
        statement {}
      }

      rate_based_statement {
        aggregate_key_type = ""
        limit = 100

        scope_down_statement {}
      }

      regex_pattern_set_reference_statement {
        arn = ""

        field_to_match {} # see above
        text_transformation {} # see above
      }

      rule_group_reference_statement {
        arn = ""

        excluded_rule {
          name = ""
        }
      }

      size_constraint_statement {
        comparison_operator = ""

        field_to_match {} # see above

        size = 1

        text_transformation {} # see above
      }

      sqli_match_statement {
        field_to_match {} # see above

        text_transformation {} # see above
      }

      xss_match_statement {
        field_to_match {} # see above

        text_transformation {} # see above
      }
    }
  }

  visibility_config {
    cloudwatch_metrics_enabled = true
    metric_name = ""
    sampled_requests_enabled = true
  }
}
```

### References

- Requires: https://github.com/terraform-providers/terraform-provider-aws/pull/11172
- API Reference: https://docs.aws.amazon.com/waf/latest/APIReference/API_Operations_AWS_WAFV2.html
- SDK Reference: https://docs.aws.amazon.com/sdk-for-go/api/service/wafv2/

### PR DESCRIPTION
**Pull request by pvanbuijtene**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #11175
Relates #11046

This PR can not be merged before #12119 #12284 are merged because of copied code for testing purpose.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
* **New Resource**: `aws_wafv2_rule_group` (#12677)
```

Output from acceptance testing:
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAwsWafv2RuleGroup'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsWafv2RuleGroup -timeout 120m
...
--- PASS: TestAccAwsWafv2RuleGroup_Disappears (117.19s)
--- PASS: TestAccAwsWafv2RuleGroup_Minimal (168.75s)
--- PASS: TestAccAwsWafv2RuleGroup_RegexPatternSetReferenceStatement (185.19s)
--- PASS: TestAccAwsWafv2RuleGroup_IpSetReferenceStatement (192.05s)
--- PASS: TestAccAwsWafv2RuleGroup_ChangeCapacityForceNew (279.87s)
--- PASS: TestAccAwsWafv2RuleGroup_ChangeNameForceNew (280.85s)
--- PASS: TestAccAwsWafv2RuleGroup_ChangeMetricNameForceNew (283.66s)
--- PASS: TestAccAwsWafv2RuleGroup_GeoMatchStatement (306.63s)
--- PASS: TestAccAwsWafv2RuleGroup_SizeConstraintStatement (307.42s)
--- PASS: TestAccAwsWafv2RuleGroup_SqliMatchStatement (309.75s)
--- PASS: TestAccAwsWafv2RuleGroup_ByteMatchStatement (310.09s)
--- PASS: TestAccAwsWafv2RuleGroup_XssMatchStatement (311.26s)
--- PASS: TestAccAwsWafv2RuleGroup_Basic (313.17s)
--- PASS: TestAccAwsWafv2RuleGroup_Tags (334.43s)
--- PASS: TestAccAwsWafv2RuleGroup_RuleAction (352.86s)
--- PASS: TestAccAwsWafv2RuleGroup_LogicalRuleStatements (355.66s)
--- PASS: TestAccAwsWafv2RuleGroup_ByteMatchStatement_FieldToMatch (475.34s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       443.877s
...
```
There's a lot of code required for testing `IpSetRefStatement` and `RegexPatternSetRefStatement`, this was the only way I could make it work, maybe/hope there's an other way to simplify it.

The code for `aws_wafv2_ip_set` and `aws_wafv2_regex_pattern_set` is included for the tests, this should be removed after their PRs are merged.

The field `statement` is recursive this is implemented with a maximum of 3 levels, more levels had a serious performance impact resulting in long test runs. I don't know if this impact is only during testing or if it also impacts run time.

### CODE REVIEW

**Comment by anGie44** [comment]
hi @pvanbuijtene 👋 -- dropping a note here to rebase and isolate the resource changes when you get a chance :)

**Comment by pvanbuijtene** [comment]
@anGie44 it's rebased.
With "isolate the resource changes" you mean to remove the duplicated code for testing purposes?

File: aws/resource_aws_wafv2_rule_group.go [hunk 1]
```diff
@@ -0,0 +1,1361 @@
...
+
+	if err := d.Set("rule", flattenWafv2Rules(resp.RuleGroup.Rules)); err != nil {
+		return fmt.Errorf("Error setting rule: %s", err)
+	}
+
+	if err := d.Set("visibility_config", flattenWafv2VisibilityConfig(resp.RuleGroup.VisibilityConfig)); err != nil {
+		return fmt.Errorf("Error setting visibility_config: %s", err)
+	}
+
+	tags, err := keyvaluetags.Wafv2ListTags(conn, *resp.RuleGroup.ARN)
```
Comment (anGie44) [inline aws/resource_aws_wafv2_rule_group.go:203]: we should use the AWS SDK's conversion func when passing the ARN

File: aws/resource_aws_wafv2_rule_group.go [hunk 2]
```diff
@@ -0,0 +1,1361 @@
...
+	if err != nil {
+		return fmt.Errorf("Error listing tags for WAFv2 RuleGroup (%s): %s", *resp.RuleGroup.ARN, err)
```
Comment (anGie44) [inline aws/resource_aws_wafv2_rule_group.go:205]: same comment as above

File: aws/resource_aws_wafv2_rule_group.go [hunk 3]
```diff
@@ -0,0 +1,1361 @@
...
+						Schema: map[string]*schema.Schema{
+							"name": {
+								Type:     schema.TypeString,
+								Required: true,
+								ValidateFunc: validation.All(
+									validation.StringLenBetween(1, 40),
+									// The value is returned in lower case by the API.
+									// Trying to solve it with StateFunc and/or DiffSuppressFunc resulted in hash problem of the rule field or didn't work.
+									validateWafv2StringIsLowerCase(),
+								),
```
Comment (anGie44) [inline aws/resource_aws_wafv2_rule_group.go:557]: though, if the validation is needed here for case, we should use the `validation` package's `StringMatch` like
```
validation.StringMatch(regexp.MustCompile(`^[a-z]+$`), "must contain only lowercase alpha characters"),
```

File: aws/resource_aws_wafv2_rule_group.go [hunk 4]
```diff
@@ -0,0 +1,1361 @@
...
+					Elem: &schema.Resource{
+						Schema: map[string]*schema.Schema{
+							"name": {
+								Type:     schema.TypeString,
+								Required: true,
+								ValidateFunc: validation.All(
+									validation.StringLenBetween(1, 40),
+									// The value is returned in lower case by the API.
+									// Trying to solve it with StateFunc and/or DiffSuppressFunc resulted in hash problem of the rule field or didn't work.
+									validateWafv2StringIsLowerCase(),
```
Comment (anGie44) [inline aws/resource_aws_wafv2_rule_group.go:556]: what type of `DiffSuppressFunc` did you experiment with @pvanbuijtene ? I was imagining something like this might do the trick:
```
DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
				return strings.EqualFold(old, new)
},
```

File: aws/resource_aws_wafv2_rule_group.go [hunk 5]
```diff
@@ -0,0 +1,1361 @@
...
+						Schema: map[string]*schema.Schema{
+							"name": {
+								Type:     schema.TypeString,
+								Required: true,
+								ValidateFunc: validation.All(
+									validation.StringLenBetween(1, 30),
+									// The value is returned in lower case by the API.
+									// Trying to solve it with StateFunc and/or DiffSuppressFunc resulted in hash problem of the rule field or didn't work.
+									validateWafv2StringIsLowerCase(),
+								),
```
Comment (anGie44) [inline aws/resource_aws_wafv2_rule_group.go:576]: same comment as above

**Comment by anGie44** [comment]
@pvanbuijtene thanks again for this PR! do you mind also adding a test for the `Empty` case where `rule` is an empty block?

File: aws/resource_aws_wafv2_rule_group_test.go [hunk 6]
```diff
@@ -0,0 +1,2486 @@
...
+			{
+				Config: testAccAwsWafv2RuleGroupConfig_Minimal(ruleGroupName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsWafv2RuleGroupExists(resourceName, &v),
+					testAccMatchResourceAttrRegionalARN(resourceName, "arn", "wafv2", regexp.MustCompile(`regional/rulegroup/.+$`)),
+					resource.TestCheckResourceAttr(resourceName, "capacity", "2"),
+					resource.TestCheckResourceAttr(resourceName, "name", ruleGroupName),
+					resource.TestCheckResourceAttr(resourceName, "description", ""),
+					resource.TestCheckResourceAttr(resourceName, "scope", wafv2.ScopeRegional),
+					resource.TestCheckResourceAttr(resourceName, "visibility_config.#", "1"),
```
Comment (anGie44) [inline aws/resource_aws_wafv2_rule_group_test.go:613]: let's also verify the `rule.#` here

File: aws/resource_aws_wafv2_rule_group_test.go [hunk 7]
```diff
@@ -0,0 +1,2486 @@
...
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+				ImportStateIdFunc: testAccAwsWafv2RuleGroupImportStateIdFunc(resourceName),
+			},
+		},
+	})
+}
+
+func TestAccAwsWafv2RuleGroup_Minimal(t *testing.T) {
```
Comment (anGie44) [inline aws/resource_aws_wafv2_rule_group_test.go:594]: could we similarly add a minimal test where `rule` is provided as the empty block `rule {}` ?

File: aws/resource_aws_wafv2_rule_group_test.go [hunk 8]
```diff
@@ -0,0 +1,2486 @@
...
+					resource.TestCheckResourceAttr(resourceName, "name", ruleGroupName),
+					resource.TestCheckResourceAttr(resourceName, "description", ruleGroupName),
+					resource.TestCheckResourceAttr(resourceName, "scope", wafv2.ScopeRegional),
+					resource.TestCheckResourceAttr(resourceName, "visibility_config.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "visibility_config.0.cloudwatch_metrics_enabled", "false"),
+					resource.TestCheckResourceAttr(resourceName, "visibility_config.0.metric_name", "friendly-metric-name"),
+					resource.TestCheckResourceAttr(resourceName, "visibility_config.0.sampled_requests_enabled", "false"),
+					resource.TestCheckResourceAttr(resourceName, "tags.%", "2"),
+					resource.TestCheckResourceAttr(resourceName, "tags.Tag1", "Value1"),
+					resource.TestCheckResourceAttr(resourceName, "tags.Tag2", "Value2"),
```
Comment (anGie44) [inline aws/resource_aws_wafv2_rule_group_test.go:41]: let's also verify the rule.# in these Checks as we're updating from 0 -> 2

File: aws/resource_aws_wafv2_rule_group.go (hunk 4 above)
Comment (pvanbuijtene) [reply aws/resource_aws_wafv2_rule_group.go:556]: @anGie44  I tried it with your suggestion, but the test still results in: `After applying this step and refreshing, the plan was not empty`, I'll attach the output. I tried to get it to work, but couldn't and my suspicion is that it's related to the calculated hash index with the lower and upper case.


I would be happy with any suggestion to get this to work without the lower-case validation :)

Comment (pvanbuijtene) [reply aws/resource_aws_wafv2_rule_group.go:556]: I can't attach a file for some reason, this is the test output:
```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 1 -run=TestAccAwsWafv2RuleGroup_ByteMatchStatement_FieldToMatch -timeout 120m
=== RUN   TestAccAwsWafv2RuleGroup_ByteMatchStatement_FieldToMatch
=== PAUSE TestAccAwsWafv2RuleGroup_ByteMatchStatement_FieldToMatch
=== CONT  TestAccAwsWafv2RuleGroup_ByteMatchStatement_FieldToMatch
--- FAIL: TestAccAwsWafv2RuleGroup_ByteMatchStatement_FieldToMatch (42.55s)
    testing.go:683: Step 0 error: After applying this step and refreshing, the plan was not empty:

        DIFF:

        DESTROY/CREATE: aws_wafv2_rule_group.test
          arn:                                                                                "arn:aws:wafv2:eu-central-1:342992114854:regional/rulegroup/rule-group-nitea/5f7ebcac-bb9d-49b8-88b9-73ff2210f243" => "<computed>"
          capacity:                                                                           "15" => "15"
          description:                                                                        "" => ""
          id:                                                                                 "5f7ebcac-bb9d-49b8-88b9-73ff2210f243" => "<computed>"
          lock_token:                                                                         "a49fb90e-02c3-4056-8866-529b6473bf45" => "<computed>"
          name:                                                                               "rule-group-nitea" => "rule-group-nitea"
          rule.#:                                                                             "1" => "1" (forces new resource)
          rule.0.action.#:                                                                    "1" => "1" (forces new resource)
          rule.0.action.0.allow.#:                                                            "1" => "1" (forces new resource)
          rule.0.action.0.block.#:                                                            "0" => "0" (forces new resource)
          rule.0.action.0.count.#:                                                            "0" => "0" (forces new resource)
          rule.0.name:                                                                        "rule-1" => "rule-1" (forces new resource)
          rule.0.priority:                                                                    "1" => "1" (forces new resource)
          rule.0.statement.#:                                                                 "1" => "1" (forces new resource)
          rule.0.statement.0.and_statement.#:                                                 "0" => "0" (forces new resource)
          rule.0.statement.0.byte_match_statement.#:                                          "1" => "1" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.#:                         "1" => "1" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.all_query_arguments.#:   "0" => "0" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.body.#:                  "0" => "0" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.method.#:                "0" => "0" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.query_string.#:          "0" => "0" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.single_header.#:         "1" => "1" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.single_header.0.name:    "a-forty-character-long-header-name-40-40" => "A-forty-character-long-header-name-40-40" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.single_query_argument.#: "0" => "0" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.uri_path.#:              "0" => "0" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.positional_constraint:                    "CONTAINS" => "CONTAINS" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.search_string:                            "word" => "word" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.text_transformation.#:                    "1" => "1" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.text_transformation.0.priority:           "1" => "1" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.text_transformation.0.type:               "NONE" => "NONE" (forces new resource)
          rule.0.statement.0.geo_match_statement.#:                                           "0" => "0" (forces new resource)
          rule.0.statement.0.ip_set_reference_statement.#:                                    "0" => "0" (forces new resource)
          rule.0.statement.0.not_statement.#:                                                 "0" => "0" (forces new resource)
          rule.0.statement.0.or_statement.#:                                                  "0" => "0" (forces new resource)
          rule.0.statement.0.regex_pattern_set_reference_statement.#:                         "0" => "0" (forces new resource)
          rule.0.statement.0.size_constraint_statement.#:                                     "0" => "0" (forces new resource)
          rule.0.statement.0.sqli_match_statement.#:                                          "0" => "0" (forces new resource)
          rule.0.statement.0.xss_match_statement.#:                                           "0" => "0" (forces new resource)
          rule.0.visibility_config.#:                                                         "1" => "1" (forces new resource)
          rule.0.visibility_config.0.cloudwatch_metrics_enabled:                              "false" => "false" (forces new resource)
          rule.0.visibility_config.0.metric_name:                                             "friendly-rule-metric-name" => "friendly-rule-metric-name" (forces new resource)
          rule.0.visibility_config.0.sampled_requests_enabled:                                "false" => "false" (forces new resource)
          scope:                                                                              "REGIONAL" => "REGIONAL"
          visibility_config.#:                                                                "1" => "1"
          visibility_config.0.cloudwatch_metrics_enabled:                                     "false" => "false"
          visibility_config.0.metric_name:                                                    "friendly-metric-name" => "friendly-metric-name"
          visibility_config.0.sampled_requests_enabled:                                       "false" => "false"



        STATE:

        aws_wafv2_rule_group.test:
          ID = 5f7ebcac-bb9d-49b8-88b9-73ff2210f243
          provider = provider.aws
          arn = arn:aws:wafv2:eu-central-1:342992114854:regional/rulegroup/rule-group-nitea/5f7ebcac-bb9d-49b8-88b9-73ff2210f243
          capacity = 15
          description =
          lock_token = a49fb90e-02c3-4056-8866-529b6473bf45
          name = rule-group-nitea
          rule.# = 1
          rule.0.action.# = 1
          rule.0.action.0.allow.# = 1
          rule.0.name = rule-1
          rule.0.priority = 1
          rule.0.statement.# = 1
          rule.0.statement.0.byte_match_statement.# = 1
          rule.0.statement.0.byte_match_statement.0.field_to_match.# = 1
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.single_header.# = 1
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.single_header.0.name = a-forty-character-long-header-name-40-40
          rule.0.statement.0.byte_match_statement.0.positional_constraint = CONTAINS
          rule.0.statement.0.byte_match_statement.0.search_string = word
          rule.0.statement.0.byte_match_statement.0.text_transformation.# = 1
          rule.0.statement.0.byte_match_statement.0.text_transformation.0.priority = 1
          rule.0.statement.0.byte_match_statement.0.text_transformation.0.type = NONE
          rule.0.visibility_config.# = 1
          rule.0.visibility_config.0.cloudwatch_metrics_enabled = false
          rule.0.visibility_config.0.metric_name = friendly-rule-metric-name
          rule.0.visibility_config.0.sampled_requests_enabled = false
          scope = REGIONAL
          visibility_config.# = 1
          visibility_config.0.cloudwatch_metrics_enabled = false
          visibility_config.0.metric_name = friendly-metric-name
          visibility_config.0.sampled_requests_enabled = false
FAIL
FAIL    github.com/terraform-providers/terraform-provider-aws/aws       43.060s
FAIL
make: *** [GNUmakefile:26: testacc] Error 1

```

File: aws/resource_aws_wafv2_rule_group_test.go (hunk 7 above)
Comment (pvanbuijtene) [reply aws/resource_aws_wafv2_rule_group_test.go:594]: The empty `rule {}` block results in missing required fields, when added the test is similar as the `Basic` test 🤔

File: aws/resource_aws_wafv2_rule_group.go (hunk 4 above)
Comment (anGie44) [reply aws/resource_aws_wafv2_rule_group.go]: ah i see, no worries thanks for the output! going to try on my end and see if maybe it's because the value is quite nested in that statement struct

Comment (anGie44) [reply aws/resource_aws_wafv2_rule_group.go]: I see what you were saying regarding the hash index 😢 perhaps this `StateFunc` schema attribute could be the alternative https://www.terraform.io/docs/extend/schemas/schema-behaviors.html#statefunc we seek

Comment (pvanbuijtene) [reply aws/resource_aws_wafv2_rule_group.go]: >  but a hash function
I was trying to prevent this, as it is quite some cumbersome work 😬 

Is this not a bug (missing feature) in the terraform code ?

Comment (pvanbuijtene) [reply aws/resource_aws_wafv2_rule_group.go]: Using the `StateFunc` results in other "weird" behavior `panic: missing field in set: rule.486868203.statement.0.byte_match_statement.0.field_to_match.0.single_header.0.name.486868203` it can't find the field anymore 🤔 

I think the current schema code is not entirely prepared for a schema like this with it's API requirements/behavior.

I'm not sure if using a hash func on the rule level is the proper solution for some problem with a field this deep in the schema, can @bflad help with some guidance here 😬

File: aws/resource_aws_wafv2_rule_group_test.go (hunk 7 above)
Comment (anGie44) [reply aws/resource_aws_wafv2_rule_group_test.go:600]: ah of course  🤦‍♀️i overlooked the first level under `rule` was not fully optional..disregard please!

File: aws/resource_aws_wafv2_rule_group.go (hunk 4 above)
Comment (anGie44) [reply aws/resource_aws_wafv2_rule_group.go]: hi @pvanbuijtene, asking about this internally, it's a valid work-around to use the `validationfunc` in this case as suggested by the schema's nested nature and statefuncs will require adjustments further up the schema. with that being said, let's use the validation func's in place but if you don't mind can we adjust the message to include `must contain only lowercase...` ?

File: aws/resource_aws_wafv2_rule_group.go (hunk 3 above)
Comment (anGie44) [reply aws/resource_aws_wafv2_rule_group.go:558]: let's keep this validatonfunc but slightly adjust the message to include "lowercase"

File: aws/resource_aws_wafv2_rule_group.go [hunk 9]
```diff
@@ -0,0 +1,1346 @@
...
+}
+
+func flattenWafv2VisibilityConfig(config *wafv2.VisibilityConfig) interface{} {
+	if config == nil {
+		return []interface{}{}
+	}
+
+	m := map[string]interface{}{
+		"cloudwatch_metrics_enabled": aws.BoolValue(config.CloudWatchMetricsEnabled),
+		"metric_name":                aws.String(*config.MetricName),
```
Comment (anGie44) [inline aws/resource_aws_wafv2_rule_group.go:1341]: to keep everything in this map as values, let's use the string value with `aws.StringValue(config.MetricName)` instead

File: aws/resource_aws_wafv2_rule_group.go [hunk 10]
```diff
@@ -0,0 +1,1346 @@
...
+}
+
+func flattenWafv2ByteMatchStatement(b *wafv2.ByteMatchStatement) interface{} {
+	if b == nil {
+		return []interface{}{}
+	}
+
+	m := map[string]interface{}{
+		"field_to_match":        flattenWafv2FieldToMatch(b.FieldToMatch),
+		"positional_constraint": b.PositionalConstraint,
```
Comment (anGie44) [inline aws/resource_aws_wafv2_rule_group.go:1150]: similar comment as below to consistently set the values in this map; positional constraint can be set with `aws.StringValue(b.PositionalConstraint)`

File: aws/resource_aws_wafv2_rule_group_test.go [hunk 11]
```diff
@@ -0,0 +1,2488 @@
...
+				},
+			},
+			"statement": []interface{}{
+				map[string]interface{}{
+					"and_statement":        []interface{}{},
+					"byte_match_statement": []interface{}{},
+					"geo_match_statement":  []interface{}{},
+					"ip_set_reference_statement": []interface{}{
+						map[string]interface{}{
+							"arn": *r.Rules[0].Statement.IPSetReferenceStatement.ARN,
```
Comment (anGie44) [inline aws/resource_aws_wafv2_rule_group_test.go:968]: Nit: extract with `aws.StringValue()`

File: aws/resource_aws_wafv2_rule_group_test.go [hunk 12]
```diff
@@ -0,0 +1,2488 @@
...
+				map[string]interface{}{
+					"and_statement":              []interface{}{},
+					"byte_match_statement":       []interface{}{},
+					"geo_match_statement":        []interface{}{},
+					"ip_set_reference_statement": []interface{}{},
+					"not_statement":              []interface{}{},
+					"or_statement":               []interface{}{},
+					"regex_pattern_set_reference_statement": []interface{}{
+						map[string]interface{}{
+							"arn": *r.Rules[0].Statement.RegexPatternSetReferenceStatement.ARN,
```
Comment (anGie44) [inline aws/resource_aws_wafv2_rule_group_test.go:1020]: Nit: extract with aws.StringValue()

File: website/docs/r/wafv2_rule_group.html.markdown [hunk 13]
```diff
@@ -0,0 +1,466 @@
...
+* `single_query_argument` - (Optional) Inspect a single query argument. See [Single Query Argument](#single-query-argument) below for details.
+* `uri_path` - (Optional) Inspect the request URI path. This is the part of a web request that identifies a resource, for example, `/images/daily-ad.jpg`.
+
+### Single Header
+
+Inspect a single header. Provide the name of the header to inspect, for example, `User-Agent` or `Referer`. This setting isn't case sensitive.
+
+The `single_header` block supports the following arguments:
+
+* `name` - (Optional) The name of the query header to inspect. This setting isn't case sensitive, only use lower case characters.
```
Comment (anGie44) [inline website/docs/r/wafv2_rule_group.html.markdown:428]: this line may be a bit contradictory (i.e. for `name` to not be case sensitive but only lower case). we can re-word this to say something like "This setting must be provided as lower case characters.")

File: website/docs/r/wafv2_rule_group.html.markdown [hunk 14]
```diff
@@ -0,0 +1,466 @@
...
+* `body` - (Optional) Inspect the request body, which immediately follows the request headers.
+* `method` - (Optional) Inspect the HTTP method. The method indicates the type of operation that the request is asking the origin to perform.
+* `query_string` - (Optional) Inspect the query string. This is the part of a URL that appears after a `?` character, if any.
+* `single_header` - (Optional) Inspect a single header. See [Single Header](#single-header) below for details.
+* `single_query_argument` - (Optional) Inspect a single query argument. See [Single Query Argument](#single-query-argument) below for details.
+* `uri_path` - (Optional) Inspect the request URI path. This is the part of a web request that identifies a resource, for example, `/images/daily-ad.jpg`.
+
+### Single Header
+
+Inspect a single header. Provide the name of the header to inspect, for example, `User-Agent` or `Referer`. This setting isn't case sensitive.
```
Comment (anGie44) [inline website/docs/r/wafv2_rule_group.html.markdown:424]: these examples could be misleading (while completely valid per AWS docs) to a reader w/o further reading of the bullet points below - to err on the side of caution, we could append parens with something like: "for example, User-Agent or Referer(provided as lowercase strings)."

File: aws/resource_aws_wafv2_rule_group.go (hunk 3 above)
Comment (anGie44) [reply aws/resource_aws_wafv2_rule_group.go:558]: @pvanbuijtene last note ^^ to adjust the error message that will appear in the output

File: aws/resource_aws_wafv2_rule_group.go (hunk 5 above)
Comment (anGie44) [reply aws/resource_aws_wafv2_rule_group.go:577]: @pvanbuijtene last note ^^ to adjust the error message that will appear in the output

**Comment by pvanbuijtene** [comment]
> Just some minor comments re: validation messages 

@anGie44  missed those, thanks for the reminder

**Comment by anGie44** [comment]
hi @pvanbuijtene,  i think the messaging is almost there 😅  just that we still need reference to the number and hyphen characters as you had it previously: `must contain only lowercase alphanumeric characters, underscores, and hyphens`

**Comment by pvanbuijtene** [comment]
@anGie44 went a bit to fast there, it's corrected.

**Comment by anGie44** [comment]
no worries @pvanbuijtene, thank you for the update!
